### PR TITLE
Ignore convert nodes from nullable to non-nullable

### LINKED ISSUE #16653: Eliminate nullability-only Convert nodes
Given the following query:

```c#
await AssertQuery<Order>(isAsync, os =>
    os.Select(o => new Order { OrderDate = o.OrderDate.Value }));
```

we get the following tree (since OrderDate is nullable):

```
value(Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1[Microsoft.EntityFrameworkCore.TestModels.Northwind.Order]).Select(o => new Order() {OrderDate = Convert(o.OrderDate.Value, Nullable`1)})
```

which makes us generate the following over-complicated SQL:

```sql
SELECT CAST(o."OrderDate" AS timestamp without time zone) AS "OrderDate" FROM "Orders" AS o
```

FWIW I think the old pipeline didn't have this useless cast.

### PR DESCRIPTION
**Pull request by roji**
Fixes #16653

### CODE REVIEW

**Comment by roji** [comment]
Note: I put a simple test directly in SqlServer, because the only point here is to assert on the SQL. I can move to specifications or eliminate entirely.

**Comment by smitpatel** [comment]
Move to specs and assert SQL

File: test/EFCore.SqlServer.FunctionalTests/Query/SimpleQuerySqlServerTest.cs [hunk 1]
```diff
@@ -4877,6 +4877,19 @@ SELECT COUNT(*)
     WHERE (([c].[CustomerID] = [o].[CustomerID]) AND [o].[CustomerID] IS NOT NULL) AND (([c].[CustomerID] = [o].[CustomerID]) AND [o].[CustomerID] IS NOT NULL)) > 0");
         }
 
+
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual async Task Convert_node_from_nullable_is_ignored(bool isAsync)
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/SimpleQuerySqlServerTest.cs]: Convert_to_nullable_on_nullable_value_is_ignored.

**Comment by roji** [comment]
Done

**Comment by roji** [comment]
Flaky test failure, merging:

```
      Microsoft.Data.SqlClient.SqlException : Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server 
is not responding.
      ---- System.ComponentModel.Win32Exception : The wait operation timed out.
      Stack Trace:
           at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
           at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncC
lose)
           at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSe
t bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
```

**Comment by smitpatel** [comment]
Re-run the test for flaky test in future.

**Pull request merged**